doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,82 +1,181 @@
-            
-    """
-    Removes the input node from its current spot in the 
-    List and inserts it as the new head node of the List.
-    """
-    def move_to_front(self, node):
-        pass
-        
-    """
-    Removes the input node from its current spot in the 
-    List and inserts it as the new tail node of the List.
-    """
-    def move_to_end(self, node):
-        pass
-
-    """
-    Deletes the input node from the List, preserving the 
-    order of the other elements of the List.
-    """
-    def delete(self, node):
-        pass
-
-    """
-    Finds and returns the maximum value of all the nodes 
-    in the List.
-    """
-    def get_max(self):
-        pass
\ No newline at end of file
+"""
+Each ListNode holds a reference to its previous node
+as well as its next node in the List.
+"""
+class ListNode:
+    def __init__(self, value, prev=None, next=None):
+        self.prev = prev
+        self.value = value
+        self.next = next
+    
+    def delete(self):
+        if self.prev:
+            self.next.prev = self.prev
+        if self.next:
+            self.prev.next = self.next
+            
+"""
+Our doubly-linked list class. It holds references to 
+the list's head and tail nodes.
+"""
+class DoublyLinkedList:
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+    
+    """
+    Wraps the given value in a ListNode and inserts it 
+    as the new head of the list. Don't forget to handle 
+    the old head node's previous pointer accordingly.
+    """
+    def add_to_head(self, value):
+        #create new_node
+        new_node = ListNode(value)
+        if self.head is None: #adding to an empty list
+            self.head = new_node
+            self.tail = new_node
+        
+        else:#adding to non empty list
+             # head -> tail -> None    newNode 
+            new_node.next = self.head # newNode -> head
+            self.head.prev = new_node # newNode   <- head
+            self.head = new_node # head = newNode
+            # newHead -> oldHead -> tail -> None
+        self.length += 1
+        
+    """
+    Removes the List's current head node, making the
+    current head's next node the new head of the List.
+    Returns the value of the removed Node.
+    """
+    def remove_from_head(self):
+        old_head = self.head
+        if self.head == None:# remove head from empty list
+            return None   
+        if self.head is self.tail:
+            self.head = None
+            self.tail = None 
+        elif self.tail.prev is self.head:
+            old_head = self.tail.prev
+            self.tail.prev = None # head -> tail <-> None
+            self.head.next = None
+            self.head = self.tail
+        else:
+            self.head.next = None # head  tail <-> None
+        self.length -= 1
+        return old_head.value

```

Comment:
This is a good start! You're reasoning about the different possible cases and how to manipulate the pointers in each case.

But we only need to check if the list is empty, and if there's only one node in the list which you're handling already on Line 57 and 59.

In every other case, to remove the head, we can just have `self.head` point to the next node, then have `self.head.prev` set to None:

```
    def remove_from_head(self):
        old_head = self.head
        if self.head == None:
            return None
        if self.head is self.tail:
            self.head = None
            self.tail = None
        else:
            # have head point to the next node in the list
            self.head = self.head.next
            # then remove any reference to the prev head by setting it to None 
            self.head.prev = None
        self.length -= 1
        return old_head.value
```



You can save yourself a bit of trouble here however, because the `LinkedList.delete` method handles all of the logic for you. So you can replace line 57 through 69 with one line: `self.delete(self.head)`:

```
    def remove_from_head(self):
        old_head = self.head
        self.delete(self.head)
        return old_head.value
```